packages/semcom-upload-component/lib/components/sem-com-register.machine.ts

```diff
@@ -0,0 +1,337 @@
+    || !event.uploadFormContext.labelInput
+    || !event.uploadFormContext.description
+    || !event.uploadFormContext.author
+    || !event.uploadFormContext.tag
+    || !event.uploadFormContext.shapes
+    || !event.uploadFormContext.version
+    || event.uploadFormContext.latest === undefined
+    || !event.uploadFormContext.checksum) {
+
+    return of(new DataNotSavedEvent());
+
+  }
+
+  // create the quad of the component
+  let data = `<#semcom-component>
+  a <http://semcom.digita.ai/voc#component>;
+  <http://semcom.digita.ai/voc#uri> <${event.uploadFormContext.uri}>;
+  <http://semcom.digita.ai/voc#label> "${event.uploadFormContext.labelInput}";
+  <http://semcom.digita.ai/voc#description> "${event.uploadFormContext.description}";
+  <http://semcom.digita.ai/voc#author> "${event.uploadFormContext.author}";
+  <http://semcom.digita.ai/voc#tag> "${event.uploadFormContext.tag}";
+  <http://semcom.digita.ai/voc#shape> `;
+
+  // shapes can contain multiple strings, so add each of them separated by a comma
+  for (const url of event.uploadFormContext.shapes.split(',')) {
+
+    data += `<${url}>,`;
+
+  }
+
+  // replace the last comma from the shapes with a ';'
+  data = data.substr(0, data.length-1);
+
+  data += `;
+  <http://semcom.digita.ai/voc#version> "${event.uploadFormContext.version}";
+  <http://semcom.digita.ai/voc#latest> "${event.uploadFormContext.latest}";
+  <http://semcom.digita.ai/voc#checksum> "${event.uploadFormContext.checksum}".`;
+
+  // Hash the component's uri to serve as the file name
+  const shaObj = new jsSHA('SHAKE256', 'TEXT');
+  shaObj.update(event.uploadFormContext.uri);
+  const hash = shaObj.getHash('B64', { outputLen: 256 }).replace(/[+=/]/g, '').toLowerCase();
+
+  // fetch the contents of the pod
+  return from(solidFetch(context.url, { headers: { 'Accept': 'text/turtle' } })).pipe(
+    switchMap((response) => zip(of(response), from(response.text()))),
+    // return a set of all filenames within the pod
+    switchMap(([ response, responseText ]) => of(new Set(
+      response.status === 200
+        ? new Parser({ format: 'Turtle' }).parse(responseText)
+          .filter((quad) => quad.object.value === 'http://www.w3.org/ns/ldp#Resource')
+          .filter((quad) => quad.subject.value !== '')
+          .map((quad) => quad.subject.value)
+        : undefined
+    ))),
+    // if the pod is offline, return an empty set
+    catchError(() => of(new Set<string>())),
+    switchMap((components) => {
+
+      // if the component already contains a component with the same name, dont save the data
+      if (components.has(hash)) { return of(new DataNotSavedEvent()); } else {
+
+        // save the metadata of the component to the pod, with the hashed url as the filename
+        return from(solidFetch(context.url, {
+          method: 'POST',
+          headers: {
+            'slug': hash,
+            'content-type': 'text/turtle',
+          },
+          body: data,
+        })).pipe(
```

Comment:
Probably no nested pipe needed. For example, you could split the map in two, one fetching and one sending the event, and throw-and-catch if the hash already is present.